File: framework/lib/property_action.dart
part of uxml;

/**
 * Implements a property setter declaration for effects
 *
 * @author:ferhat@ (Ferhat Buyukkokten)
 */
class PropertyAction extends Action {

  /** Name of target element to animate */
  String target;

  /**
   * Property to set.
   */
  Object property;

  /**
   * Target value of property.
   */
  Object value;

  /**
   * Delay in ms before action is executed.
   */
  int delay = 0;

  PropertyAction() : super() {
  }

  /**
   * Starts the action by scheduling an animation task.
   */
  void start(UxmlElement chromeRoot, UxmlElement control) {
    ActionData data = getActionData(control);
    if (data.actionState != ActionData.ACTION_ACTIVE) {
      data.actionState = ActionData.ACTION_ACTIVE;
      if (delay != 0) {
        Application.current.scheduler.schedule(delay,
            0, _delayedExecute, this, [chromeRoot, control]);
      } else {
        _applyActionValue(chromeRoot, control);
      }
    }
  }

  void _delayedExecute(num tweenValue, Object tag){
    List<Object> tagParts = tag;
    UxmlElement chromeRoot = tagParts[0];
    UxmlElement control = tagParts[1];
    ActionData data = getActionData(control);
    _applyActionValue(chromeRoot, control);
    data.actionState = ActionData.ACTION_IDLE;
  }

  void _applyActionValue(UxmlElement chromeRoot, UxmlElement control) {
    ActionData data = getActionData(control);
    if (target == null) {
      // If target was not specified, the owner itself is the target
      data.targetElement = chromeRoot;
      if (!(data.targetElement is UIElement)) {
        throw new Exception("Invalid targetElement");
      }
    } else {
      // resolve target name. Either id or id.subObject
      if (chromeRoot is UIElement) {
        UIElement uiChromeRoot = chromeRoot;
        if (target.indexOf(".", 0) == -1) {
            data.targetElement = uiChromeRoot.getElement(target);
            if (!(data.targetElement is UxmlElement)) {
              data.targetElement = null;
            }
        } else {
          List<String> targetChain = target.split(".");
          UxmlElement element = uiChromeRoot.getElement(targetChain[0]);
          for (int i = 1; i < targetChain.length; i++) {
            PropertyDefinition propDef = PropertyCache.getPropertyDefinition(
                element.getDefinition(), targetChain[i]);
            element = element.getProperty(propDef);
          }
          data.targetElement = element;
          if (!(data.targetElement is IKeyValue)) {
            throw new Exception("Invalid targetElement");
          }
        }
      }
    }
    if (data.targetElement != null) {
      data.startValue = data.targetElement.getProperty(property);
      Action.applyPropertyValue(this, data.targetElement, property, value);
    }
  }

  /**
   * Resets action to original idle state before starting action or
   * once reversal is complete.
   */
  void reset(UxmlElement control) {
    ActionData data = getActionData(control);
    data.actionState = ActionData.ACTION_IDLE;
  }

  /**
   * Reverses the action.
   */
  void reverse(UxmlElement control) {
    ActionData data = getActionData(control);
    if (data.actionState == ActionData.ACTION_ACTIVE &&
        data.targetElement != null) {
      reversePropertyValue(this, data.targetElement, property);
      data.actionState = ActionData.ACTION_IDLE;
    }
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/effects/PropertyAction.java
package com.hello.uxml.tools.framework.effects;

import com.hello.uxml.tools.framework.PropertyDefinition;
import com.hello.uxml.tools.framework.UIElement;
import com.hello.uxml.tools.framework.UxmlElement;

/**
 * Implements a property setter declaration for effects
 *
 * @author ferhat
 */
public class PropertyAction extends Action {
  private String target;
  private PropertyDefinition property;
  private Object value;
  private int delay;

  /**
   * Constructor.
   */
  public PropertyAction() {
  }

  /**
   * Sets or returns target element name.
   */
  public String getTarget() {
    return  target;
  }

  public void setTarget(String target) {
    this.target = target;
  }

  /**
   * Sets or returns property definition.
   */
  public PropertyDefinition getProperty() {
    return property;
  }

  public void setProperty(PropertyDefinition property) {
    this.property = property;
  }

  /**
   * Sets or return target value of action.
   */
  public Object getValue() {
    return value;
  }

  public void setValue(Object value) {
    this.value = value;
  }

  /**
   * Sets or return delay time in ms of action.
   */
  public int getDelay() {
    return delay;
  }

  public void setDelay(int delay) {
    this.delay = delay;
  }

  /**
  * Starts the action.
  */
  @Override
  public void start(UxmlElement chromeRoot, UxmlElement control) {
    ActionData data = getActionData(control);
    // TODO(ferhat): implement delay
    if (data.getActionState() != ActionData.ACTION_ACTIVE) {
      data.setTargetElement((target != null && chromeRoot instanceof UIElement) ?
          ((UIElement) control).getElement(target) : control);
      data.setStartValue(data.getTargetElement().getProperty(property));
      data.setActionState(ActionData.ACTION_ACTIVE);
      data.getTargetElement().setProperty(property, value);
    }
  }

  /**
  * Reverses the action.
  */
  @Override
  public void reverse(UxmlElement control) {
    ActionData data = getActionData(control);
    if (data.getActionState() == ActionData.ACTION_ACTIVE) {
      data.getTargetElement().setProperty(property, data.getStartValue());
      data.setActionState(ActionData.ACTION_IDLE);
    }
  }
}
